### DISCUSSION TITLE
Pass Referer in Datasource Proxy

### DISCUSSION THREAD

**Discussion by disfluxly** [grafana/grafana#38367] [Dashboards, Panels & Visualizations]
**What would you like to be added**:
Allow passing the `Referer` header in the data source proxy.

**Why is this needed**:
As part of https://github.com/grafana/grafana/issues/13949, the `Referer` was removed from getting sent along with the data source proxy. This was highly useful for troubleshooting purposes.

For example, this allowed us to pinpoint which dashboards were causing performance problems due to bad queries. This was _pivotal_ in helping us quickly identify the source of long-running queries that would cause the CPU on Influx to spike.

We currently have a different proxy that sits between Grafana and our data sources to do introspection of the queries and route them accordingly. We wanted to use the `Referer` in order to tag our Jaeger Traces. This would allow us to very easily, and in close to realtime, identify queries that took (for example) 5 seconds to execute and which dashboard executed that query.

**Comment by marefr** [comment]
There's already X-Dashboard-Id and X-Panel-Id headers sent (I think) that you could use. Issue open for adding X-Dashboard-Uid (#17410). Could possibly add an X-Referer header?

**Comment by disfluxly** [comment]
@marefr 

Hmmm. I'm not seeing those two headers when I'm inspecting request headers outbound from the UI.

I also don't see them in the headers my application receives from the dataproxy.

A quick search in the repo shows that they may only be available for graphite/prometheus datasources. I'm using Influx.

**Comment by Mugglephilip** [comment]
We also need this feature, since we use the Referer to count each dashboard's visits.

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by judilsteve** [comment]
Bumping this. We recently upgraded from 5.x to 9.1.2 and the `Referer` header is gone in the new version; we were using this to identify poorly optimised dashboards that were flooding our datasource (OpenTSDB) with queries. We also are not seeing `X-Dashboard-Id` or `X-Panel-Id` being sent from the frontend; only `x-grafana-org-id` (which is useless info to us as we have only one org ID).

Essentially we now have no way of knowing which dashboards are generating the bulk of our queries; whereas this information was logged by previous versions.

**Comment by marefr** [comment]
#60040 was just merged. From Grafana v9.4.0 the `X-Grafana-Referer` will be forwarded and contain the original `Referer` value.

**Reply by madhu-reddy-peram** [reply]
@marefr - X-Grafana-Referer header still doesn't come for `api/ds/query` calls from grafana for prometheus datasource. Do you have any issue opened to address this already. This header is useful for metrics and logs queries to block bad dashbaords.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
